From ac6aa30ac47b2abaf142f76de46374da2a98f6e7 Mon Sep 17 00:00:00 2001
From: Matthew Rosato <mjrosato@linux.ibm.com>
Date: Fri, 3 Dec 2021 09:27:06 -0500
Subject: [PATCH] s390x/pci: add supported DT information to clp response

The DTSM is a mask that specifies which I/O Address Translation designation
types are supported.  Today QEMU only supports DT=1.

Signed-off-by: Matthew Rosato <mjrosato@linux.ibm.com>
Reviewed-by: Eric Farman <farman@linux.ibm.com>
Reviewed-by: Pierre Morel <pmorel@linux.ibm.com>
Message-Id: <20211203142706.427279-5-mjrosato@linux.ibm.com>
Signed-off-by: Thomas Huth <thuth@redhat.com>

Origin: upstream, https://gitlab.com/qemu-project/qemu/commit/ac6aa30ac4
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1853307
Last-Update: 2023-07-04

---
 hw/s390x/s390-pci-bus.c         | 1 +
 hw/s390x/s390-pci-inst.c        | 1 +
 hw/s390x/s390-pci-vfio.c        | 1 +
 include/hw/s390x/s390-pci-bus.h | 1 +
 include/hw/s390x/s390-pci-clp.h | 3 ++-
 5 files changed, 6 insertions(+), 1 deletion(-)

Index: qemu/hw/s390x/s390-pci-bus.c
===================================================================
--- qemu.orig/hw/s390x/s390-pci-bus.c	2023-07-04 20:38:39.000000000 -0400
+++ qemu/hw/s390x/s390-pci-bus.c	2023-07-04 20:51:02.000000000 -0400
@@ -782,6 +782,7 @@
     resgrp->i = 128;
     resgrp->maxstbl = 128;
     resgrp->version = 0;
+    resgrp->dtsm = ZPCI_DTSM;
 }
 
 static void set_pbdev_info(S390PCIBusDevice *pbdev)
Index: qemu/hw/s390x/s390-pci-inst.c
===================================================================
--- qemu.orig/hw/s390x/s390-pci-inst.c	2023-07-04 20:38:39.000000000 -0400
+++ qemu/hw/s390x/s390-pci-inst.c	2023-07-04 20:51:02.000000000 -0400
@@ -329,6 +329,7 @@
         stw_p(&resgrp->i, group->zpci_group.i);
         stw_p(&resgrp->maxstbl, group->zpci_group.maxstbl);
         resgrp->version = group->zpci_group.version;
+        resgrp->dtsm = group->zpci_group.dtsm;
         stw_p(&resgrp->hdr.rsp, CLP_RC_OK);
         break;
     }
Index: qemu/hw/s390x/s390-pci-vfio.c
===================================================================
--- qemu.orig/hw/s390x/s390-pci-vfio.c	2023-07-04 20:38:39.000000000 -0400
+++ qemu/hw/s390x/s390-pci-vfio.c	2023-07-04 20:51:03.000000000 -0400
@@ -160,6 +160,7 @@
         resgrp->i = cap->noi;
         resgrp->maxstbl = cap->maxstbl;
         resgrp->version = cap->version;
+        resgrp->dtsm = ZPCI_DTSM;
     }
 }
 
Index: qemu/include/hw/s390x/s390-pci-bus.h
===================================================================
--- qemu.orig/include/hw/s390x/s390-pci-bus.h	2023-07-04 20:49:59.000000000 -0400
+++ qemu/include/hw/s390x/s390-pci-bus.h	2023-07-04 20:51:02.000000000 -0400
@@ -37,6 +37,7 @@
 #define ZPCI_MAX_UID 0xffff
 #define UID_UNDEFINED 0
 #define UID_CHECKING_ENABLED 0x01
+#define ZPCI_DTSM 0x40
 
 OBJECT_DECLARE_SIMPLE_TYPE(S390pciState, S390_PCI_HOST_BRIDGE)
 OBJECT_DECLARE_SIMPLE_TYPE(S390PCIBus, S390_PCI_BUS)
Index: qemu/include/hw/s390x/s390-pci-clp.h
===================================================================
--- qemu.orig/include/hw/s390x/s390-pci-clp.h	2023-07-04 20:38:39.000000000 -0400
+++ qemu/include/hw/s390x/s390-pci-clp.h	2023-07-04 20:49:59.000000000 -0400
@@ -163,7 +163,8 @@
     uint8_t fr;
     uint16_t maxstbl;
     uint16_t mui;
-    uint64_t reserved3;
+    uint8_t dtsm;
+    uint8_t reserved3[7];
     uint64_t dasm; /* dma address space mask */
     uint64_t msia; /* MSI address */
     uint64_t reserved4;
